oml/datasets/audios.py

```diff
@@ -0,0 +1,353 @@
+        self.num_frames = int(num_seconds * sr)
+        self.num_channels = num_channels
+
+        if extra_data is not None:
+            assert all(
+                len(record) == len(paths) for record in extra_data.values()
+            ), "All the extra records need to have the size equal to the dataset's size"
+            self.extra_data = extra_data
+        else:
+            self.extra_data = {}
+
+        self.frame_offsets: List[int] = self.extra_data.get(input_frame_offset_key)
+
+        self.input_tensors_key = input_tensors_key
+        self.index_key = index_key
+
+        self._paths = [Path(p) if dataset_root is None else Path(dataset_root) / p for p in paths]
+
+    def _downmix_and_resample(self, audio: FloatTensor, sr: int) -> FloatTensor:
+        """
+        Downmix audio to mono and resample it to the dataset's sampling rate.
+
+        Args:
+            audio (FloatTensor): Input audio tensor.
+            sr (int): Original sampling rate of the audio.
+
+        Returns:
+            FloatTensor: Processed audio tensor.
+        """
+        if audio.shape[0] != self.num_channels:
+            audio = audio.mean(dim=1, keepdim=True)
+        if sr != self.sr:
+            resampler = Resample(sr, self.sr)
+            audio = resampler(audio)
+        return audio
+
+    def _trim_or_pad(self, audio: FloatTensor, frame_offset: int) -> FloatTensor:
+        """
+        Trim or pad the audio to match the desired number of frames.
+
+        Args:
+            audio (FloatTensor): Audio tensor.
+            frame_offset (int): Starting frame offset for trimming.
+
+        Returns:
+            FloatTensor: Trimmed or padded audio tensor.
+        """
+        audio_length = audio.shape[1]
+        if audio_length > self.num_frames:
+            if frame_offset is None:
+                frame_offset = random.randrange(0, audio_length - self.num_frames)
+            audio = audio[:, frame_offset : frame_offset + self.num_frames]
+        else:
+            padding = (self.num_frames - audio_length, 0)
+            audio = torch.nn.functional.pad(audio, padding)
+        return audio
+
+    def get_audio(self, item: int) -> FloatTensor:
+        """
+        Load and process an audio file.
+
+        Args:
+            item (int): Dataset item index.
+
+        Returns:
+            FloatTensor: Processed audio tensor.
+        """
+        path = self._paths[item]
+        assert (
+            path.suffix in AUDIO_EXTENSIONS
+        ), f"Input audio has invalid extension '{path.suffix}'. It should be one of {AUDIO_EXTENSIONS}."
+        audio, sr = torchaudio.load(path)
+        frame_offset = None if self.frame_offsets is None else self.frame_offsets[item]
+        audio = self._downmix_and_resample(audio, sr)
+        audio = self._trim_or_pad(audio, frame_offset)
+        return audio
+
+    def get_spectral_repr(self, audio: FloatTensor, params: Dict[str, Any] = DEFAULT_MELSPEC_PARAMS) -> FloatTensor:
```

Comment:
do we need this method as public as far as we have two `visualize` methods which look very similar?
moreover, this method is almost static which may be confusing

i suggest to make the method private

Reply:
the idea is to have this method as public, so a user could override it with different spectral representation extraction method (e.g. spectrogram/CQT/MFCC/...) for visualisation

Reply:
Suppose, I train and validate a model with [CQTs](https://en.wikipedia.org/wiki/Constant-Q_transform) (which looks a bit different from melspectrograms), and when validating via retrieval I see melspectrograms instead of CQTs and can not gain insights why model works in such way. 
So, I need to rewrite the visualisation method. 

And why then do we need the private method if it won't be (potentially) used?

Reply:
As an idea, I can remove this method from class and move it to `utils`. Then, I'll add the additional argument (e.g. `spec_repr_func: Callable[[FloatTensor], FloatTensor]`) in `visualize/visualize_with_player` methods.

something like this:
`datasets/audios`
```
from utils.audios import default_spec_repr_func
class Dataset(...):
...
    def visualize(
        self,
        item: int,
        color: TColor = BLACK,
        spec_repr_func: Callable[[FloatTensor], FloatTensor] = default_spec_repr_func,
    ) -> NDArray[np.uint8]:
        """
        Visualize an audio file.

        Args:
            item: Dataset item index.
            color: Color of the plot.
            spec_repr_func: Spectral representation extraction function

        Returns:
            Array representing the image of the plot.
        """
        audio = self.get_audio(item)
        spec_repr = spec_repr_func(audio).squeeze(0)
        return visualize_audio(spec_repr=spec_repr, color=color)
...
``` 
`utils/audios`
```
def default_spec_repr_func(audio: FloatTensor) -> FloatTensor:
    """
    Generate a log-scaled Spectrogram from an audio signal.
    Used primarily for visualization.

    Args:
        audio: The input audio tensor.

    Returns:
        The spectral representation of the input audio tensor.
    """
    spec = Spectrogram(n_fft=1024)(audio)
    log_spec = torch.log1p(spec).squeeze(0)
    return log_spec
```

Reply:
Okay, let's do this way: `spec_repr_func: Callable[[FloatTensor], FloatTensor] = default_spec_repr_func,`

Reply:
done